File: app/src/main/java/com/netspeed/monitor/SpeedMonitorService.java
package com.netspeed.monitor;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Intent;
import android.content.pm.ServiceInfo;
import android.graphics.Bitmap;
import android.graphics.drawable.Icon;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;

import java.util.Timer;
import java.util.TimerTask;

/**
 * Foreground service that monitors network speed every 100ms
 * and updates a persistent status bar notification with live speed values.
 * UI callbacks fire at 100ms for responsive gauges; notification updates
 * are throttled to every 500ms to avoid excessive system overhead.
 */
public class SpeedMonitorService extends Service {

    // Intent action constants for starting and stopping the service
    public static final String ACTION_START = "com.netspeed.monitor.action.START";
    public static final String ACTION_STOP = "com.netspeed.monitor.action.STOP";

    // Unique notification ID for the persistent foreground notification
    private static final int NOTIFICATION_ID = 1001;

    // Speed calculation interval in milliseconds (100ms = 10 updates/sec)
    private static final long SPEED_INTERVAL_MS = 100;
    // Notification update interval: update every Nth tick to reduce system overhead
    // 5 ticks × 100ms = 500ms between notification updates
    private static final int NOTIFICATION_THROTTLE_TICKS = 5;

    // Timer that fires speed calculation every 100ms
    private Timer timer;
    // Tick counter for throttling notification updates
    private int tickCount;
    // Handler for posting UI callbacks from the timer thread to the main thread
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    // Reference to the notification manager for posting updates
    private NotificationManager notificationManager;
    // The shared traffic stats calculator
    private TrafficStatsCalculator calculator;

    // Static flag indicating whether the service is currently running
    private static volatile boolean running;

    /**
     * Callback interface for delivering speed updates to the UI.
     */
    public interface SpeedCallback {
        void onSpeedUpdate(double download, double upload, long sessionRx, long sessionTx);
    }

    // Static callback reference; the Activity registers/unregisters as listener
    private static volatile SpeedCallback callback;

    /** Registers a callback to receive live speed updates from the service. */
    public static void setCallback(SpeedCallback cb) {
        callback = cb;
    }

    /** Returns true if the service timer is actively running. */
    public static boolean isRunning() {
        return running;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        notificationManager = getSystemService(NotificationManager.class);
        calculator = NetSpeedApp.getCalculator();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null) {
            if (ACTION_START.equals(intent.getAction())) {
                // Start foreground monitoring
                startMonitoring();
            } else if (ACTION_STOP.equals(intent.getAction())) {
                // Stop monitoring and shut down the service
                stopMonitoring();
            }
        }
        // START_STICKY: system restarts the service if it's killed
        return START_STICKY;
    }

    @Override
    public IBinder onBind(Intent intent) {
        // No binding supported
        return null;
    }

    /**
     * Promotes to foreground with an initial notification and starts the speed timer.
     * Guards against double-start: cancels any existing timer before creating a new one.
     */
    private void startMonitoring() {
        // Cancel any existing timer to prevent duplicate scheduling on rapid taps
        if (timer != null) {
            timer.cancel();
            timer = null;
        }

        // Reset calculator for a fresh session
        NetSpeedApp.resetCalculator();
        calculator = NetSpeedApp.getCalculator();

        // Build initial zero-speed notification to satisfy foreground requirement
        Notification notification = buildNotification("0 B/s", "0 B/s", null);

        // Start foreground with service type for API 29+ compatibility
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            startForeground(NOTIFICATION_ID, notification,
                    ServiceInfo.FOREGROUND_SERVICE_TYPE_DATA_SYNC);
        } else {
            startForeground(NOTIFICATION_ID, notification);
        }

        running = true;
        // Reset tick counter for notification throttling
        tickCount = 0;

        // Refresh home screen widgets to show "Live" status
        SpeedWidgetProvider.refreshAllWidgets(this);

        // Schedule speed calculation every 100ms for responsive UI updates
        timer = new Timer("SpeedTimer", true);
        timer.scheduleAtFixedRate(new TimerTask() {
            @Override
            public void run() {
                // Calculate current speed on the timer thread
                double[] speeds = calculator.calculateSpeed();
                long[] session = calculator.getSessionBytes();
                double dl = speeds[0];
                double ul = speeds[1];
                long sRx = session[0];
                long sTx = session[1];

                // Throttle notification updates to every 500ms (every 5th tick)
                // Notifications are heavyweight; updating too often wastes CPU and battery
                tickCount++;
                if (tickCount >= NOTIFICATION_THROTTLE_TICKS) {
                    tickCount = 0;

                    // Format speed for notification text
                    String dlText = SpeedUtils.formatSpeed(dl);
                    String ulText = SpeedUtils.formatSpeed(ul);

                    // Generate dynamic bitmap icon showing speed in status bar
                    Bitmap iconBitmap = SpeedIconGenerator.createSpeedIcon(dl);
                    Icon icon = Icon.createWithBitmap(iconBitmap);

                    // Update the notification with latest speed values
                    Notification n = buildNotification(dlText, ulText, icon);
                    notificationManager.notify(NOTIFICATION_ID, n);

                    // Broadcast speed to home screen widget (same 500ms throttle)
                    Intent widgetIntent = new Intent(SpeedWidgetProvider.ACTION_SPEED_UPDATE);
                    widgetIntent.setPackage(getPackageName());
                    widgetIntent.putExtra(SpeedWidgetProvider.EXTRA_DOWNLOAD_SPEED, dl);
                    widgetIntent.putExtra(SpeedWidgetProvider.EXTRA_UPLOAD_SPEED, ul);
                    sendBroadcast(widgetIntent);
                }

                // Deliver speed update to the UI callback on the main thread (every 100ms)
                mainHandler.post(() -> {
                    SpeedCallback cb = callback;
                    if (cb != null) {
                        cb.onSpeedUpdate(dl, ul, sRx, sTx);
                    }
                });
            }
        }, 0, SPEED_INTERVAL_MS);
    }

    /**
     * Stops the timer, removes the foreground notification, and terminates the service.
     */
    private void stopMonitoring() {
        running = false;
        if (timer != null) {
            timer.cancel();
            timer = null;
        }
        stopForeground(STOP_FOREGROUND_REMOVE);

        // Refresh home screen widgets to show "Stopped" status and reset speed
        SpeedWidgetProvider.refreshAllWidgets(this);

        stopSelf();
    }

    /**
     * Builds a Notification with download/upload speed displayed.
     * Uses a dynamic bitmap icon when available, falls back to a static resource.
     */
    private Notification buildNotification(String downloadSpeed, String uploadSpeed, Icon dynamicIcon) {
        // PendingIntent to open MainActivity when notification is tapped
        PendingIntent contentIntent = PendingIntent.getActivity(
                this, 0,
                new Intent(this, MainActivity.class),
                PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_UPDATE_CURRENT
        );

        // PendingIntent for the "Stop" action button in the notification
        Intent stopIntent = new Intent(this, SpeedMonitorService.class);
        stopIntent.setAction(ACTION_STOP);
        PendingIntent stopPendingIntent = PendingIntent.getService(
                this, 1, stopIntent,
                PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_UPDATE_CURRENT
        );

        // Choose the notification icon: dynamic bitmap or static resource
        Icon smallIcon = (dynamicIcon != null)
                ? dynamicIcon
                : Icon.createWithResource(this, R.drawable.ic_speed_notification);

        // Build notification with speed values prominently displayed
        Notification.Builder builder = new Notification.Builder(this, NetSpeedApp.CHANNEL_ID)
                .setSmallIcon(smallIcon)
                .setContentTitle("\u2193 " + downloadSpeed + "   \u2191 " + uploadSpeed)
                .setContentText("Tap to open Net Speed Monitor")
                .setSubText("Net Speed")
                .setContentIntent(contentIntent)
                .setOngoing(true)
                .setOnlyAlertOnce(true)
                .setShowWhen(false)
                .setCategory(Notification.CATEGORY_SERVICE)
                .setPriority(Notification.PRIORITY_MAX)
                .setSortKey("0")
                .setBadgeIconType(Notification.BADGE_ICON_NONE)
                .setNumber(0)
                .addAction(new Notification.Action.Builder(
                        Icon.createWithResource(this, R.drawable.ic_stop),
                        getString(R.string.notification_stop),
                        stopPendingIntent
                ).build())
                .setStyle(new Notification.BigTextStyle()
                        .setBigContentTitle("\u2193 " + downloadSpeed + "   \u2191 " + uploadSpeed)
                        .bigText("Download: " + downloadSpeed + "\nUpload: " + uploadSpeed));

        // Show notification immediately on Android 12+ (no 10-second delay)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            builder.setForegroundServiceBehavior(Notification.FOREGROUND_SERVICE_IMMEDIATE);
        }

        return builder.build();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        running = false;
        // Clean up timer to prevent leaks
        if (timer != null) {
            timer.cancel();
            timer = null;
        }
    }
}

File: app/src/main/java/com/netspeed/monitor/service/SpeedMonitorService.kt
package com.netspeed.monitor.service

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Intent
import android.content.pm.ServiceInfo
import android.os.Build
import android.os.IBinder
import androidx.core.app.NotificationCompat
import androidx.core.app.ServiceCompat
import com.netspeed.monitor.MainActivity
import com.netspeed.monitor.R
import com.netspeed.monitor.domain.model.NetworkSpeed
import com.netspeed.monitor.domain.usecase.ObserveNetworkSpeedUseCase
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import javax.inject.Inject

// Foreground service that continuously monitors network speed and displays it in a sticky notification
@AndroidEntryPoint
class SpeedMonitorService : Service() {

    // Injected use case for observing network speed from the domain layer
    @Inject
    lateinit var observeNetworkSpeedUseCase: ObserveNetworkSpeedUseCase

    // Supervised coroutine scope tied to the service lifecycle; SupervisorJob prevents child failure from cancelling siblings
    private val serviceScope = CoroutineScope(Dispatchers.Default + SupervisorJob())

    // Reference to system notification manager for building and updating notifications
    private lateinit var notificationManager: NotificationManager

    companion object {
        // Unique notification channel ID for network speed updates
        const val CHANNEL_ID = "speed_monitor_channel"
        // Unique ID for the persistent foreground notification
        const val NOTIFICATION_ID = 1001
        // Action string for the start intent to begin monitoring
        const val ACTION_START = "com.netspeed.monitor.action.START"
        // Action string for the stop intent to stop monitoring
        const val ACTION_STOP = "com.netspeed.monitor.action.STOP"
    }

    // Called when the service is first created; set up notification channel and manager
    override fun onCreate() {
        super.onCreate()
        // Obtain the system notification manager service
        notificationManager = getSystemService(NotificationManager::class.java)
        // Create the notification channel required for Android 8.0+ (API 26)
        createNotificationChannel()
    }

    // Called when startService/startForegroundService is invoked with an Intent
    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        when (intent?.action) {
            // Start monitoring: promote to foreground and begin speed observation
            ACTION_START -> startMonitoring()
            // Stop monitoring: clean up and stop the service
            ACTION_STOP -> stopMonitoring()
        }
        // START_STICKY ensures the system restarts the service if it's killed
        return START_STICKY
    }

    // Services do not support binding in this app; return null
    override fun onBind(intent: Intent?): IBinder? = null

    // Promotes the service to foreground with an initial notification and begins speed collection
    private fun startMonitoring() {
        // Build an initial notification to satisfy the foreground service requirement
        val notification = buildNotification(
            downloadSpeed = "0 B/s",
            uploadSpeed = "0 B/s"
        )

        // Start as a foreground service using ServiceCompat for API compatibility
        ServiceCompat.startForeground(
            this,
            NOTIFICATION_ID,
            notification,
            // Specify DATA_SYNC type for network monitoring (required on API 34+)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q)
                ServiceInfo.FOREGROUND_SERVICE_TYPE_DATA_SYNC
            else 0
        )

        // Launch a coroutine to observe speed updates and refresh the notification
        serviceScope.launch {
            // Collect the continuous stream of NetworkSpeed values from the use case
            observeNetworkSpeedUseCase().collect { speed ->
                // Update the notification with the latest speed values
                updateNotification(speed)
            }
        }
    }

    // Cancels the coroutine scope, stops foreground promotion, and self-terminates
    private fun stopMonitoring() {
        // Cancel all running coroutines in the service scope
        serviceScope.cancel()
        // Remove the foreground notification and demote to background
        stopForeground(STOP_FOREGROUND_REMOVE)
        // Schedule service for termination
        stopSelf()
    }

    // Updates the persistent notification with the latest speed data
    private fun updateNotification(speed: NetworkSpeed) {
        // Format speed values to human-readable strings using domain model helpers
        val downloadText = NetworkSpeed.formatSpeed(speed.downloadSpeed)
        val uploadText = NetworkSpeed.formatSpeed(speed.uploadSpeed)

        // Build a new notification with updated speed values
        val notification = buildNotification(
            downloadSpeed = downloadText,
            uploadSpeed = uploadText
        )

        // Push the updated notification to the system; replaces the existing one by same ID
        notificationManager.notify(NOTIFICATION_ID, notification)
    }

    // Builds a Notification with download and upload speed prominently displayed
    private fun buildNotification(downloadSpeed: String, uploadSpeed: String): Notification {
        // PendingIntent to open MainActivity when the notification is tapped
        val pendingIntent = PendingIntent.getActivity(
            this,
            0,
            Intent(this, MainActivity::class.java),
            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
        )

        // Stop action intent for the notification action button
        val stopPendingIntent = PendingIntent.getService(
            this,
            1,
            Intent(this, SpeedMonitorService::class.java).apply { action = ACTION_STOP },
            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
        )

        // Build notification with live speed as the primary content
        return NotificationCompat.Builder(this, CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_speed_notification)
            // Speed values as the main title for maximum visibility in collapsed view
            .setContentTitle("↓ $downloadSpeed   ↑ $uploadSpeed")
            .setContentText("Tap to open Net Speed Monitor")
            // Subtext appears in the notification header area
            .setSubText("Net Speed")
            .setContentIntent(pendingIntent)
            .setOngoing(true)
            .setSilent(true)
            .setShowWhen(false)
            // Prevent re-alerting on each update
            .setOnlyAlertOnce(true)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            // Categorize as a running service notification
            .setCategory(NotificationCompat.CATEGORY_SERVICE)
            // Show notification immediately on Android 12+ (no delay)
            .setForegroundServiceBehavior(NotificationCompat.FOREGROUND_SERVICE_IMMEDIATE)
            .addAction(
                R.drawable.ic_stop,
                getString(R.string.notification_stop),
                stopPendingIntent
            )
            // Expanded view shows detailed speed breakdown
            .setStyle(
                NotificationCompat.BigTextStyle()
                    .setBigContentTitle("↓ $downloadSpeed   ↑ $uploadSpeed")
                    .bigText("Download: $downloadSpeed\nUpload: $uploadSpeed")
            )
            .build()
    }

    // Creates the notification channel required on Android 8.0 (API 26) and higher
    private fun createNotificationChannel() {
        // Channel with LOW importance: shows in shade but never makes sound
        val channel = NotificationChannel(
            CHANNEL_ID,
            "Network Speed Monitor",
            NotificationManager.IMPORTANCE_LOW
        ).apply {
            // Human-readable description shown in system notification settings
            description = "Live network upload and download speed monitoring"
            // Disable vibration for frequent update notifications
            enableVibration(false)
            // Disable sound for silent status updates
            setSound(null, null)
        }
        // Register the channel with the system notification manager
        notificationManager.createNotificationChannel(channel)
    }

    // Called when the service is destroyed; cancel coroutines to prevent leaks
    override fun onDestroy() {
        super.onDestroy()
        // Clean up the coroutine scope to stop all ongoing work
        serviceScope.cancel()
    }
}
